pkg/apis/pipeline/v1beta1/pipeline_types.go

```diff
@@ -174,7 +175,16 @@ func (pt PipelineTask) Deps() []string {
 	deps = append(deps, pt.resourceDeps()...)
 	deps = append(deps, pt.orderingDeps()...)
 
-	return deps
+	uniqueDeps := sets.NewString()
+	for _, w := range deps {
+		if uniqueDeps.Has(w) {
+			continue
+		}
+		uniqueDeps.Insert(w)
+
+	}
+
+	return uniqueDeps.List()
```

Comment:
this significantly reduces the call to `dag.visit` and makes it possible for the unit test to complete within five seconds. The mysterious dot notation in `dag.visit` is still unsolved:

https://github.com/tektoncd/pipeline/blob/747f4baa014e44a9143eafb7d87a8f9a6c5a3e60/pkg/reconciler/pipeline/dag/dag.go#L145

Its creating a link between two nodes which indirectly connected (`C.A` in `A -> B -> C`) but where/how is this data being utilized is a mystery.